[Issue-308]: Add support for reading directly from file handles

slobodan-ilic opened the pull request:
File Handle Support for SAV Files

Summary

Adds support for reading SAV files directly from file-like objects (e.g., zip archives, BytesIO) without extracting to disk.

Use Case

Read large SAV files from zip archives without temporary file extraction:

import zipfile
import pyreadstat

with zipfile.ZipFile('data.zip', 'r') as zf:
    with zf.open('survey.sav') as f:
        df, meta = pyreadstat.read_sav(f)  # No extraction needed!

Implementation

- Leverages ReadStat's built-in I/O handler system
- Custom Cython handlers bridge Python file objects to C callbacks
- Unified parser function handles both paths and file objects
- Zero breaking changes - existing code works unchanged

Testing

- All existing tests pass (92 tests)
- New integration test validates zip archive reading
- Data integrity verified (identical output to path-based reading)

Changes

- pyreadstat.read_sav() now accepts file-like objects with read() and seek() methods
- +132 net lines (after refactoring to eliminate duplication)
- Extensible to other formats (DTA, SAS7BDAT, etc.)

ofajardo committed:
Merge pull request #303 from Roche/dev

Dev

ofajardo commented:
hi, thanks a lot for the PR! It is really great! 

 Is this ready or WIP? 

I read your PR quickly, and I have a couple of questions, a couple of asks and one suggestion:

questions: 
* Why is the _read_fileobj implemented only for sav? Would it benefit also all the other reading functions?
* in read_sav, if a file object you route it to _read_fileobj, which skips apply_value_formats ... why? I think that is still needed.

suggestions:
* Would it be possible to just use the regular run_conversion instead of having a separate _read_fileobj? It looks to me as if run_conversion could be used out of the box with the changes you did. Maybe it is something you did for testing purposes but now can be discarded? This would solve my previous questions.

asks:
* Could you add in the README file in the section Usage -> more reading options, a new sub-section describing briefly this new feature, and a couple of examples on how it could be used? I am thinking on your zip usecase ,and also what is described in #279 if it applies (reading from a remote file)
* The test_read_sav_from_zip_file_handle, could you put it in test_narwhalified.py and make sure it works for both pandas and polars?

slobodan-ilic commented:
Hey @ofajardo , thanks for your comment. I think all of your observations are valid. I should've created the PR as a draft (I just converted it to one just now). I first wanna make sure our team can use it for what they need, and if it proves to be usable, then I'll get back to the PR and make it much tighter, and also address all of your concerns.

ofajardo commented:
Awesome! Thanks!

ofajardo committed:
Merge pull request #312 from Roche/dev

version 1.3.2

slobodan-ilic committed:
Add file handle support for reading from file-like objects

- Add custom I/O handlers in _readstat_parser.pyx to bridge Python file objects to ReadStat C library
- Modify run_conversion() to detect file-like objects and use custom handlers
- Works for all read functions (SAV, SAS7BDAT, DTA, XPORT, POR)
- Add test in test_basic.py and test_narwhalified.py (pandas/polars compatible)
- Add README documentation with zip archive example

Addresses maintainer feedback:
- Uses run_conversion() directly instead of separate wrapper
- Preserves apply_value_formats and all other options
- File handle support works for all formats, not just SAV

slobodan-ilic committed:
Add remote file example to README and BytesIO test

slobodan-ilic committed:
Add HTTP integration test for file handle support

Test demonstrates reading SAV files from remote sources via HTTP, showing real-world usage of file handle support. Uses built-in http.server - no external dependencies needed.

slobodan-ilic commented:
@ofajardo We finally tested this in the scope of our software, and it does what we want it to. I've updated the PR addressing all of your concerns. Do you think it's cool to put it from draft to an actual PR?

ofajardo commented:
sure please go ahead. I am a bit busy these days, but will try to review soon.

ofajardo commented:
I changed the base to the dev branch, as I would merge into there. 

I had a quick look at the changes and they look great! I am a bit confused though, because I only see  3 recent commits from a couple of days ago, and I do not see any of the changes we discussed before, for example, I do not see any change at all to pyreadstat.pyx ... 

or maybe you just discarded the previous commits as those changes were not needed anymore?

BTW I see that the signature of run_readstat_parser has changed in the pyx file but in the pxd file has not been updated.

ofajardo commented:
OK, I have reviewed now and everything looks good. I assume you just got rid of the previous commits and changes.
I also compiled and run tests, everything went smoothly.
I used to have all functions in the pyx listed in the pxd, and I see that you did not include the new ones. Just reading again about it, I see that there is no efficiency gain on having not cimported functions from another module listed in the pxd, so I guess it can stay as it is, as I also if the signature of a function changes in the pyx then you have to change it twice also in the pxd for no real performance gain. 
So, please just confirm everything is also good from your side, and I will merge and then do a release.

ofajardo commented:
@slobodan-ilic from your point of view is it ready for merging and releasing?

Nidhalios commented:
Does this enhancement allow finally to read objects (buffer) with read_sas7bdat and read_file_in_chunks ? or it requires additional dev to add such feature ?

slobodan-ilic commented:
> @slobodan-ilic from your point of view is it ready for merging and releasing?

Yes, if it passes all builds and tests - I'd say it's ready. Thanks for your support, sorry for delayed response :)

slobodan-ilic commented:
> Does this enhancement allow finally to read objects (buffer) with read_sas7bdat and read_file_in_chunks ? or it requires additional dev to add such feature ?

I haven't tested that, probably best to create a separate "integration" test like the one I added in this PR for the sav file.

ofajardo commented:
OK cool, I will merge into a feature branch to be able to run my CI/CD pipeline and if everything is good, I make the release.

ofajardo commented:
@slobodan-ilic I got to run now, but the CI/CD pipeline if failing on the zip file test you added, here the [link](https://github.com/ofajardo/pyreadstat_wheels5/actions/runs/21259108939/job/61181781729) in case you could take a look

ofajardo commented:
ok it seems I have a mix of errors coming from pandas 3.0 and in a couple of cases coming from your tests because /tmp seems not to be writable, but apparently only in some machines. Checking it.

ofajardo commented:
@slobodan-ilic  OK, NamedTempFile was raising issues on Windows, adding the argument delete_on_close=True makes it to work, but breaks the tests for python < 3.12. I have changed to writing to a temporary directory instead. Otherwise all OK. Will do a release, but first I will update the Readstat sources as a PR has been just merged now.

@Nidhalios I have added tests for reading in chunks and reading in multiprocessing with file-like objects, and yes, it works out of the box. Again, great work @slobodan-ilic !

Nidhalios commented:
@slobodan-ilic @ofajardo That's some very good news ! thank you for this new feature :-) ! Do you know when it will be released ?

ofajardo commented:
@slobodan-ilic @Nidhalios : released!

slobodan-ilic commented:
Thank you @ofajardo for your support on this! Sorry for being spotty throughout the PR.
